db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -188,6 +188,7 @@ Status DBImpl::FlushMemTableToOutputFile(
       error_handler_.SetBGError(log_io_s, BackgroundErrorReason::kFlush);
     }
   } else {
+    log_io_s.PermitUncheckedError();
```

Comment:
Rather than PermitUncheckedError, why not make the check below on line 198 and 209 check on log_io_s?  Alternatively, move the s = log_io_s outside of the if (!ok) since the two are always effectively the same.

The log_io_s should probably be merged with the io_s a few lines below.

Reply:
> Rather than PermitUncheckedError, why not make the check below on line 198 and 209 check on log_io_s? Alternatively, move the s = log_io_s outside of the if (!ok) since the two are always effectively the same.
> 
> The log_io_s should probably be merged with the io_s a few lines below.

log_io_s is intentioned to used only for io status of log, in this way, we are able to handle WAL write error and SST file write error in different logic.

Reply:
> I think `log_io_s` is kept separate from the `io_s` intentionally so that the execution result of `SyncClosedLogs()` is preserved. This is used for distinguishing between different failure cases, under the current control flow: function does not return early on failure, and there is one exit point. Consequently, errors that occur in different places along the code path need to be preserved if we want to handle errors in a single place later.

You are right, thanks for the detailed explanation and thanks for the fix.

Reply:
@zhichao-cao . I do think we can improve. Currently, there are multiple such `status` or `io_status` or `log_io_status` flying around, making the already-intricate code even more complex. Would you consider using a data structure such as
```
class ErrorReportAndHandlerContext {
  Status s;
  IOStatus io_s;
  std::string path;
  FileType type;
  ...
};
```
to record the error's information? This way, we do not have to use different variables, making the code cleaner.

Reply:
Yes, it is in the plan to refactor the error handling status, context, and reason.